src/webgpu/api/operation/queue/writeBuffer.spec.ts

```diff
@@ -1,11 +1,228 @@
+  'Uint8Array',
+  'Uint16Array',
+  'Uint32Array',
+  'Int8Array',
+  'Int16Array',
+  'Int32Array',
+  'Float32Array',
+  'Float64Array',
+] as const;
+
+type WriteBufferSignature = {
+  bufferOffset: number;
+  data: readonly number[];
+  arrayType: typeof kTypedArrays[number];
+  useArrayBuffer: boolean;
+  dataOffset?: number;
+  dataSize?: number;
+};
+
+class F extends GPUTest {
+  calculateRequiredBufferSize(writes: WriteBufferSignature[]): number {
+    let bufferSize = 0;
+    // Calculate size of final buffer
+    for (const { bufferOffset, data, arrayType, useArrayBuffer, dataOffset, dataSize } of writes) {
+      const typeBuilder = self[arrayType];
+      let requiredSize = data.length * typeBuilder.BYTES_PER_ELEMENT;
+
+      // When passing data as an ArrayBuffer, dataOffset, dataSize use byte instead of number of
+      // elements
+      const bytesPerElement = useArrayBuffer ? 1 : typeBuilder.BYTES_PER_ELEMENT;
+      if (dataOffset) {
+        requiredSize -= dataOffset * bytesPerElement;
+      }
+      if (dataSize) {
+        requiredSize = Math.min(requiredSize, dataSize * bytesPerElement);
+      }
+
+      requiredSize += bufferOffset;
+
+      if (requiredSize > bufferSize) {
+        bufferSize = requiredSize;
+      }
+    }
+    // writeBuffer requires buffers to be a multiple of 4
+    return Math.ceil(bufferSize / 4) * 4;
+  }
+
+  testWriteBuffer(...writes: WriteBufferSignature[]) {
+    const bufferSize = this.calculateRequiredBufferSize(writes);
+    const buffer = this.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+
+    // Initialize buffer to non-zero data (0xff)
+    const expectedData = new Uint8Array(range<number>(bufferSize, () => 0xff));
+    this.queue.writeBuffer(buffer, 0, expectedData);
+
+    for (const { bufferOffset, data, arrayType, useArrayBuffer, dataOffset, dataSize } of writes) {
+      const typeBuilder = self[arrayType];
+      const writeData = new typeBuilder(data);
+      this.queue.writeBuffer(
+        buffer,
+        bufferOffset,
+        useArrayBuffer ? writeData.buffer : writeData,
+        dataOffset,
+        dataSize
+      );
+      const bytesPerElement = useArrayBuffer ? 1 : typeBuilder.BYTES_PER_ELEMENT;
+      const begin = dataOffset ? dataOffset * bytesPerElement : 0;
+      expectedData.set(
+        new Uint8Array(
+          writeData.buffer.slice(begin, dataSize ? begin + dataSize * bytesPerElement : undefined)
+        ),
+        bufferOffset
+      );
+    }
+
+    this.debug(`expectedData: [${expectedData.join(', ')}]`);
+    this.expectContents(buffer, expectedData);
+  }
+}
+
+export const g = makeTestGroup(F);
+
+const kTestData = range<number>(16, i => i);
+
+g.test('array_types')
+  .desc('Tests that writeBuffer correctly handles different TypedArrays and ArrayBuffer.')
+  .cases(params().combine(poptions('arrayType', kTypedArrays)).combine(pbool('useArrayBuffer')))
```

Comment:
optional nit: Arguably redundant because I think with useArrayBuffer:true the test is the same regardless of the arrayType? Not a big deal though.